### PR TITLE
CLOUDSTACK-9351: Add ids parameter to resource listing API calls

### PR DESCRIPTION
**Pull request by nvazquez**
## General behaviour
A new parameter is added in each method, its type a list of IDs of the entity, it will be mutually exclusive with id. (Similar to <code>id</code> and <code>ids</code> parameters in <code>listVirtualMachines</code> method)

### API Methods affected
* <code>listTemplates</code>: new parameter **<code>ids</code>**, mutually exclusive with <code>id</code>
* <code>listVolumes</code>: new parameter **<code>ids</code>**, mutually exclusive with <code>id</code>
* <code>listSnapshots</code>: new parameter **<code>ids</code>**, mutually exclusive with <code>id</code>
* <code>listVMSnapshots</code>: new parameter **<code>vmsnapshotids</code>**, mutually exclusive with <code>vmsnapshotid</code>

### CODE REVIEW

**Comment by rafaelweingartner** [comment]
@nvazquez  have you tested this change?

I noticed that you are using the SearchCriteria, shouldn’t that “idIN” have to be at the POJO and mapped with that name? How is that going to be transformed into a select with “select…. Where .. id in ();”

**Comment by nvazquez** [comment]
Hi @rafaelweingartner, yes, I did test them. I provided new parameters in API calls and they retrieved objects as expected.
About the implementation, I was aware that <code>listVirtualMachines</code> had this parameter already implemented, so tried replicating the same idea, as it is in <code>QueryManagerImpl</code> in lines 792, 847, 937

**Comment by rafaelweingartner** [comment]
Ah ok, so our home made JPA takes care of that ;)
That felt a little weird to me, but now I understand why.
I would only suggest you squashing your commits into one. It is a single proposal right?

**Comment by nvazquez** [comment]
Yes, it is only this proposal. Cool, I'll squash them

File: server/src/com/cloud/api/query/QueryManagerImpl.java [hunk 1]
```diff
@@ -1731,6 +1731,17 @@
         Long zoneId = cmd.getZoneId();
         Long podId = cmd.getPodId();
 
+        List<Long> ids = null;
```
Comment (rafaelweingartner) [inline server/src/com/cloud/api/query/QueryManagerImpl.java]: Would you mind extracting lines 1734-1744 to a method? Then, you could write test cases and even create some documentation for it.

File: server/src/com/cloud/api/query/QueryManagerImpl.java [hunk 2]
```diff
@@ -1790,6 +1802,10 @@
             sc.setParameters("display", display);
         }
 
+        if (ids != null && !ids.isEmpty()) {
```
Comment (rafaelweingartner) [inline server/src/com/cloud/api/query/QueryManagerImpl.java]: What about extracting line 1805 - 1808 to a method. Those lines are the same as the ones at 3172-3175. then, you could also write test cases and add java doc for it.

File: server/src/com/cloud/api/query/QueryManagerImpl.java (hunk 1 above)
Comment (nvazquez) [inline server/src/com/cloud/api/query/QueryManagerImpl.java]: Ok, they extend from <code>com.cloud.utils.component.ManagerBase</code> but I found an issue on that, from its project (cloud-utils) there is no visibility to <code>com.cloud.exception.InvalidParameterValueException</code>
I wouldn't like to modify dependencies, what do you think? Would I use another exception?

**Comment by rafaelweingartner** [comment]
@nvazquez you can create a common super class for those class, and put it into the “cloud-server” project. Additionally, you can remove the “extends ManagerBase”, I can bet it is not needed; so, let’s try removing some unused hierarchy?

Also, can you remove the "@Local" annotation? it is not needed.

**Comment by nvazquez** [comment]
Thanks @rafaelweingartner I made a refactor based in your comments. I could remove "extends ManagerBase" from <code>QueryManagerImpl</code> but not from <code>VMSnapshotManagerImpl</code> and <code>SnapshotManagerImpl</code> so I decided to create static methods and use the new class as a helper instead of a superclass, do you think this is ok?

I couldn't find <code>@Local</code> to remove it, in which file did you mean to remove it?

**Comment by rafaelweingartner** [comment]
@nv I am sorry, I might have seen the “@Local” annotation in some other PR, and I thought I have seen on yours.

The classes VMSnapshotManagerImpl and SnapshotManagerImpl are using something from “ManagerBase”?

I am ok with the static methods if there is no way to create a class hierarchic. They are clean and well documented. I am only missing the test cases for those two new methods.

**Comment by nvazquez** [comment]
Cool, I'll working on test cases.
`VMSnapshotManagerImpl` is using `_name` variable from `ManagerBase,` and `SnapshotManagerImpl` overrides `configure`, `start` and `stop` method

**Comment by rafaelweingartner** [comment]
@nvazquez thanks for the analysis on the ManagerBase issue.
Giving that I think the best approach would be to let all of the three (3) classes extending it, and then we could create a hierarchy, instead of using static methods.

What do you think?

**Comment by nvazquez** [comment]
@rafaelweingartner I pushed unit tests for this new methods.
So the idea is that the 3 classes extend from a new class and not ManagerBase?

**Comment by rafaelweingartner** [comment]
That is it, and then this new class would extend ManagerBase.

**Comment by nvazquez** [comment]
Cool, I'll work on it

**Comment by nvazquez** [comment]
@rafaelweingartner I pushed new changes, would integration tests be needed for this pr? I have tested  api calls with new ids parameters and succeeded

**Comment by swill** [comment]
@nvazquez I have added this to my CI queue.  I have a pretty serious backlog, so it may take me a few days to get to it though.  In the mean time, lets see if we can get some LGTM code reviews.  :)

**Comment by nvazquez** [comment]
That's cool! Thanks @swill!

File: server/test/com/cloud/api/query/MutualExclusiveIdsManagerBaseTest.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+import com.cloud.utils.db.SearchCriteria;
+
+@RunWith(MockitoJUnitRunner.class)
+public class MutualExclusiveIdsManagerBaseTest {
+
+    @Mock
+    SearchCriteria<String> sc;
+
+    private static Long id1 = Long.valueOf(1l);
```
Comment (rafaelweingartner) [inline server/test/com/cloud/api/query/MutualExclusiveIdsManagerBaseTest.java]: here you could just use 1L.

File: server/test/com/cloud/api/query/MutualExclusiveIdsManagerBaseTest.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        idsList = Arrays.asList(id1, id2);
+        idsEmptyList = Arrays.asList();
+        expectedListId = Arrays.asList(id1);
+        expectedListIds = Arrays.asList(id1, id2);
+    }
+
+    @Test
+    public void testSetIdsListToSearchCriteria(){
+        mgr.setIdsListToSearchCriteria(sc, idsList);
+        Mockito.verify(sc, times(1)).setParameters(Mockito.same("idIN"), Mockito.same(id1), Mockito.same(id2));
```
Comment (rafaelweingartner) [inline server/test/com/cloud/api/query/MutualExclusiveIdsManagerBaseTest.java]: the default of Mockito.verify already uses "times(1)"; so, there is no need to inform it.

File: server/test/com/cloud/api/query/MutualExclusiveIdsManagerBaseTest.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    @Test
+    public void testSetIdsListToSearchCriteriaEmptyList(){
+        mgr.setIdsListToSearchCriteria(sc, idsEmptyList);
+        Mockito.verify(sc, never()).setParameters(Mockito.anyString(), Mockito.any());
+    }
+
+    @Test
+    public void testGetIdsListId(){
+        List<Long> result = mgr.getIdsListFromCmd(id1, idsEmptyList);
+        assertEquals(result, expectedListId);
```
Comment (rafaelweingartner) [inline server/test/com/cloud/api/query/MutualExclusiveIdsManagerBaseTest.java]: This is more a question of philosophy, but the assertEquals should be: assertEquals(expected, actual). You are doing the opposite. It does not mean that the way it is coded now is wrong, it is just not the normal way (at least that I am used).

**Comment by rafaelweingartner** [comment]
@nvazquez I have a question about the method "getIdsListFromCmd".
If we inform an ID and a list of IDs (not empty) at the same time, what should be the behaviors? Should we merge them in that list of IDs? Should we only use the ID? Or should we only use the IDs list?

**Comment by nvazquez** [comment]
@rafaelweingartner Actually the idea was that they were mutually exclusive, that only of them should be provided, emulating `listVirtualMachines` API method behaviour in which if both are not null, it fails. 
In general cases I think a good idea show be merging them, but in this case I think it should emulate this behaviour, maybe renaming the method to a more descriptive name?

**Comment by rafaelweingartner** [comment]
The method name is ok. I just asked to confirm how you see the method workings. I only feel that the Javadoc should say that, whenever the ID variable is not null, it is the only one that is going to be used and returned in the List. Also, we should have a test for that.

**Comment by nvazquez** [comment]
Ok, I would modify javadoc and add a test for that particular case

**Comment by rafaelweingartner** [comment]
Sorry if I am very picky sometimes; but, I believe that the more clear we are about a method behavior and the more tests to enforce that behavior we have, the better it is.

**Comment by nvazquez** [comment]
No problem, I appreciate your comments and think they are really helpful ;)

**Comment by nvazquez** [comment]
I included last changes and squashed all commits into one

**Comment by rafaelweingartner** [comment]
That is great.
Now reading your documentation, I noticed another question.
And if both are null? Or if ID is null and IDs is empty? The way it is now, it will return a null or an empty list value.
Is that the expected behavior? If so, it deserves a test case.

**Comment by nvazquez** [comment]
Cool, I added both test cases

**Comment by rafaelweingartner** [comment]
@Nv, that is great. The code now LGTM. 
There is a problem with Jenkins, maybe a push –f can solve it?
Your work as always is very good ;)
I believe now we would need to run the integration tests.

**Comment by koushik-das** [comment]
@nvazquez Can you add integration tests for all these APIs? Checkout test_deploy_vm_multiple test in test/integration/smoke/test_vm_life_cycle.py

**Comment by rhtyd** [comment]
Nice PR, though this would need some regression tests. Can you add marvin smoke tests around list APis (add them to travis yml file)

tag:needlove

**Comment by nvazquez** [comment]
Sure @rhtyd, I'll add them. I was working on it based on @koushik-das suggestion but I still have to work on it.

File: test/integration/smoke/test_list_ids_parameter.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                                   domainid=cls.account.domainid,
+                                   diskofferingid=cls.disk_offering.id
+                                   )
+            cls.debug("Created a volume with ID: %s" % volume.id)
+            cls.volumes.append(volume)
+
+        cls._cleanup = [
+                        cls.disk_offering,
+                        cls.account
+                        #,cls.volumes[0],
```
Comment (nvazquez) [inline test/integration/smoke/test_list_ids_parameter.py]: I pushed this commented lines to ask about this, if I add them test fails but I don't understand how they are deleted. After test finishes I see volumes are deleted

Comment (rafaelweingartner) [inline test/integration/smoke/test_list_ids_parameter.py]: @nvazquez , 
Sorry to let you down. I am terrible with python.
I believe @rhtyd,  @swill,  or @DaanHoogland can help you with that.

**Comment by nvazquez** [comment]
@koushik-das @rhtyd I finished writing marvin test for this feature, I copy my test results:

````
[root@ussarlabcsmgt41 cloudstack]# nosetests --with-marvin --marvin-config=setup/dev/advanced.cfg test/integration/smoke/test_list_ids_parameter.py

==== Marvin Init Started ====

=== Marvin Parse Config Successful ===

=== Marvin Setting TestData Successful===

==== Log Folder Path: /tmp//MarvinLogs//May_06_2016_09_34_04_NHXZGI. All logs will be available here ====

=== Marvin Init Logging Successful===

==== Marvin Init Successful ====
===final results are now copied to: /tmp//MarvinLogs/test_list_ids_parameter_TRKKN0===
````
````
[root@ussarlabcsmgt41 cloudstack]# cat /tmp//MarvinLogs/test_list_ids_parameter_TRKKN0/results.txt
Test listing Volumes using 'ids' parameter ... === TestName: test_01_list_volumes | Status : SUCCESS ===
ok
Test listing Templates using 'ids' parameter ... === TestName: test_02_list_templates | Status : SUCCESS ===
ok
Test listing Snapshots using 'ids' parameter ... === TestName: test_03_list_snapshots | Status : SUCCESS ===
ok
Test listing VMSnapshots using 'vmsnapshotids' parameter ... === TestName: test_04_list_vm_snapshots | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 4 tests in 1187.175s

OK

````

**Comment by koushik-das** [comment]
@nvazquez The test results show that it took ~20mins to complete. Why list tests are taking so long?
Some observations:
- Tests are not tagged, please tag them (refer to existing tests). I ran them in a simulator setup and they passed so these can be added to Travis once the overall time is reduced and proper tags added
- Templates created as part of tests is not cleaned up.
- time.sleep(720). I saw the comment about vm snapshot issue, better to disable the vm snapshot test and enable it once snapshot issue is fixed

File: test/integration/smoke/test_list_ids_parameter.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+                                cls.services["virtual_machine"],
+                                templateid=template.id,
+                                accountid=cls.account.name,
+                                domainid=cls.account.domainid,
+                                serviceofferingid=cls.service_offering.id,
+                                mode=cls.services["mode"]
+                                )
+
+        #Sleep for 12 minutes after vms creation because of bug while taking snapshots. 
+        time.sleep(720)
```
Comment (koushik-das) [inline test/integration/smoke/test_list_ids_parameter.py]: Please remove sleep and disable vm snapshot tests for now.

File: test/integration/smoke/test_list_ids_parameter.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    @classmethod
+    def tearDownClass(cls):
+        cls.apiclient = super(TestListIdsParams, cls).getClsTestClient().getApiClient()
+        try:
+            cleanup_resources(cls.apiclient, cls._cleanup)
+        except Exception as e:
+            raise Exception("Warning: Exception during cleanup : %s" % e)
+        return
+
+    def test_01_list_volumes(self):
```
Comment (koushik-das) [inline test/integration/smoke/test_list_ids_parameter.py]: Add tags as appropriate

**Comment by nvazquez** [comment]
Hi @koushik-das thanks for reviewing! Actually it took so long due to <code>setUpClass</code> method, after that, test are really simple and quick. I included 12 minutes sleep as you noticed due to vm snapshots bug, remaining time was (vm, templates, snapshots and vmsnapshots) creation

**Comment by swill** [comment]
### CI RESULTS

```
Tests Run: 85
  Skipped: 0
   Failed: 0
   Errors: 0
 Duration: 4h 38m 35s
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_09_2016_07_17_28_RFHUY6:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/DeployDataCenter__May_09_2016_07_17_28_RFHUY6/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/DeployDataCenter__May_09_2016_07_17_28_RFHUY6/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/DeployDataCenter__May_09_2016_07_17_28_RFHUY6/runinfo.txt)

**`/tmp/MarvinLogs/test_network_DL9JLX:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/test_network_DL9JLX/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/test_network_DL9JLX/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/test_network_DL9JLX/runinfo.txt)

**`/tmp/MarvinLogs/test_suite_9BKAZW:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/test_suite_9BKAZW/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/test_suite_9BKAZW/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/test_suite_9BKAZW/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_MX5M6P:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/test_vpc_routers_MX5M6P/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/test_vpc_routers_MX5M6P/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1497/tmp/MarvinLogs/test_vpc_routers_MX5M6P/runinfo.txt)


Uploads will be available until `2016-07-09 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

**Comment by swill** [comment]
I am missing one code review for this one.  @nvazquez can you re-push to kick off Travis again?  Thx...

**Comment by swill** [comment]
@nvazquez: strange, this did not run any tests.  Any ideas on that one?

```
nosetests --with-marvin --marvin-config=${marvinCfg} -s -a tags=advanced smoke/test_list_ids_parameter.py
```

**Comment by nvazquez** [comment]
Sure @swill I pushed again. I also ran again tests in my environment:

````
[root@ussarlabcsmgt41 cloudstack]# nosetests --with-marvin --marvin-config=setup/dev/advanced.cfg test/integration/smoke/test_list_ids_parameter.py

==== Marvin Init Started ====

=== Marvin Parse Config Successful ===

=== Marvin Setting TestData Successful===

==== Log Folder Path: /tmp//MarvinLogs//May_09_2016_17_39_25_ETC1VH. All logs will be available here ====

=== Marvin Init Logging Successful===

==== Marvin Init Successful ====
===final results are now copied to: /tmp//MarvinLogs/test_list_ids_parameter_A7VS1H===
[root@ussarlabcsmgt41 cloudstack]# cat /tmp//MarvinLogs/test_list_ids_parameter_A7VS1H/results.txt
Test listing Volumes using 'ids' parameter ... === TestName: test_01_list_volumes | Status : SUCCESS ===
ok
Test listing Templates using 'ids' parameter ... === TestName: test_02_list_templates | Status : SUCCESS ===
ok
Test listing Snapshots using 'ids' parameter ... === TestName: test_03_list_snapshots | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 3 tests in 446.512s

OK
````

**Comment by swill** [comment]
I am a bit confused why the attributes tag of `advanced` did not work in my test and zero tests ran. I will look at that quickly tomorrow to see if there is something different about these tests from the others I am running this way. I think this one is in pretty good shape now. Will verify everything when I am back at a computer.

**Comment by koushik-das** [comment]
@swill As I had mentioned in my last comment, newly added tests are not tagged. The command you used to run these tests has tags, don't pass any tags in order to run them.
@nvazquez Please tag the tests appropriately, something like 
@attr(tags = ["devcloud", "advanced", "advancedns", "smoke", "basic", "sg"], required_hardware="false")

**Comment by nvazquez** [comment]
@koushik-das actually I've added tags, removed time.sleep and commented out test_04

**Comment by koushik-das** [comment]
@nvazquez My bad, didn't look at the changes. Thanks for the fixing the tests.

**Comment by nvazquez** [comment]
No problem ;)

**Comment by koushik-das** [comment]
LGTM, ran the new tests.

Test listing Volumes using 'ids' parameter ... === TestName: test_01_list_volumes | Status : SUCCESS ===
ok
Test listing Templates using 'ids' parameter ... === TestName: test_02_list_templates | Status : SUCCESS ===
ok
Test listing Snapshots using 'ids' parameter ... === TestName: test_03_list_snapshots | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 3 tests in 148.845s

OK

**Comment by swill** [comment]
Sorry, apparently I am not getting enough sleep now days.  I looked at your code after you updated it to add the tags, which is why I was confused why the tests were not run in my case.  I thought I had checked that they had tags when I added them to my tests, but I blame on that on being tired too.  :)

This one is ready to merge.  Thanks guys...

**Comment by nvazquez** [comment]
Thanks @koushik-das @swill for testing! I'll push -f to restart Travis test which failed

**Comment by nvazquez** [comment]
Done, now it passed but somehow Jenkins test was not executed

**Comment by swill** [comment]
@nvazquez this seems to have broken travis for everyone.  We have this issue showing up in pretty much all travis runs right now `ContextSuite context=TestListIdsParams>:teardown`.

Can you investigate?  Thanks...

**Comment by serg38** [comment]
@swill Last Travis run for this PR was a success for test_list_ids_parameter. Which PR gives an issue?

==== Marvin Init Successful ====
=== TestName: test_01_list_volumes | Status : SUCCESS ===
=== TestName: test_02_list_templates | Status : SUCCESS ===
=== TestName: test_03_list_snapshots | Status : SUCCESS ===
===final results are now copied to: /tmp//MarvinLogs/test_list_ids_parameter_IPS7T4===

**Comment by nvazquez** [comment]
@swill it's strange, as @serg38 says it succeeds for us and for @koushik-das, I'll investigate. Is there any log of the failure?

**Comment by swill** [comment]
@serg38 I had a bunch of people re-push PRs last night and this morning to get everything green and 3 or 4 are failing with this.  

Here are a couple off the top of my head:  #1502, #1376

**Comment by serg38** [comment]
The issue that it couldn't find a snaphsot because it is already cleaned by account cleanup.
 +        cls._cleanup = [
 +                        cls.disk_offering,
 +                        cls.account,
 +                        cls.service_offering,
 +                        cls.snapshot_1,
 +                        cls.snapshot_2,
 +                        cls.snapshot_3
 +                        ]
 +

Snapshot cleanup can be safely removed since account cleanup will remove them automatically.

**Comment by swill** [comment]
Can you create a new PR to fix this and we will get it tested right away.  Thanks for being on top of this.  👍

**Comment by nvazquez** [comment]
Done @swill @serg38

**Comment by swill** [comment]
Thank you, I will merge it assuming Jenkins and Travis come back clean.  Thx... 👍

**Pull request merged**